Support for #define

The Cake script

```c#
#define SOME_SYMBOL
```

Causes:

> error CS1032: Cannot define/undefine preprocessor symbols after first token in file

(Version 0.21.1+Branch.main.Sha.9ff85bb53ea1890fabeb709d38ba79d3bf5cb16f)

Is there a technical reason this can't work or am I the first to try?

I have a real use-case now: avoiding `warning CS0162: Unreachable code detected` when I use a `const bool` instead of preprocessor.

Would you take a PR for this?

> If it's fresh in your mind, would you do it? That would be perhaps more efficient.
@jnm2 even if that were true, it still takes some effort to do it and time is limited 😎 But if no one takes it on I might take a stab at it when I can, have a few outstanding PRs awaiting @cake-build/cake-team review and could potentially look at this post that. But other things might also take precedence.

@mholo65 
So tested plain scripting  and csi.exe
```csharp
#if (FOO)
#r ".\Newtonsoft.Json.dll"
#endif


Console.WriteLine(
    Newtonsoft.Json.JsonConvert.SerializeObject(
            new {
                Hello = "World"
            }
        )
    );
```

and it works regardless if i define FOO or not, so would seem as `#r` is processed at same scope as `#if/#define`
and gives `error CS0103: The name 'Newtonsoft' does not exist in the current context` if i remove the line.

I propose we first align with what Roslyn / csi.exe does as that's likely a fairly simple fix and later on look if we can / should improve it in any way.

I do like the `DefineConstants` as cmdline/config/environment idea though - essentially just a cake config for constants, makes a lot of sense.

--- a/src/Cake.Core.Tests/Unit/Scripting/Analysis/ScriptAnalyzerResultTests.cs
+++ b/src/Cake.Core.Tests/Unit/Scripting/Analysis/ScriptAnalyzerResultTests.cs
@@ -153,6 +153,25 @@ namespace Cake.Core.Tests.Unit.Scripting.Analysis
                 Assert.Contains("using static System.Math;", result.UsingStaticDirectives);
                 Assert.Contains("using static System.Console;", result.UsingStaticDirectives);
             }
+
+            [Fact]
+            public void Should_Populate_Defines_From_Provided_Script()
+            {
+                // Given
+                var script = new ScriptInformation("./build.cake");
+                script.Defines.Add("#define FOO");
+                var other = new ScriptInformation("./other.cake");
+                other.Defines.Add("#define BAR");
+                script.Includes.Add(other);
+
+                // When
+                var result = new ScriptAnalyzerResult(script, new List<string>());
+
+                // Then
+                Assert.Equal(2, result.Defines.Count);
+                Assert.Contains("#define FOO", result.Defines);
+                Assert.Contains("#define BAR", result.Defines);
+            }
         }
     }
 }

--- a/src/Cake.Core.Tests/Unit/Scripting/Analysis/ScriptAnalyzerTests.cs
+++ b/src/Cake.Core.Tests/Unit/Scripting/Analysis/ScriptAnalyzerTests.cs
@@ -384,6 +384,21 @@ namespace Cake.Core.Tests.Unit.Scripting.Analysis
                 Assert.Equal(1, result.Script.UsingStaticDirectives.Count);
                 Assert.Equal("using static System.Math;", result.Script.UsingStaticDirectives.ElementAt(0));
             }
+
+            [Fact]
+            public void Should_Process_Define_Directives()
+            {
+                // Given
+                var fixture = new ScriptAnalyzerFixture();
+                fixture.GivenScriptExist("/Working/script.cake", "#define FOO");
+
+                // When
+                var result = fixture.Analyze("/Working/script.cake");
+
+                // Then
+                Assert.Equal(1, result.Script.Defines.Count);
+                Assert.Equal("#define FOO", result.Script.Defines.ElementAt(0));
+            }
         }
     }
 }

--- a/src/Cake.Core.Tests/Unit/Scripting/ScriptTests.cs
+++ b/src/Cake.Core.Tests/Unit/Scripting/ScriptTests.cs
@@ -15,7 +15,7 @@ namespace Cake.Core.Tests.Unit.Scripting
             public void Should_Not_Throw_If_Namespaces_Are_Null()
             {
                 // Given, When
-                var script = new Script(null, new string[] { }, new ScriptAlias[] { }, new string[] { }, new string[] { });
+                var script = new Script(null, new string[] { }, new ScriptAlias[] { }, new string[] { }, new string[] { }, new string[] { });
 
                 // Then
                 Assert.Equal(0, script.Namespaces.Count);
@@ -25,7 +25,7 @@ namespace Cake.Core.Tests.Unit.Scripting
             public void Should_Not_Throw_If_Lines_Are_Null()
             {
                 // Given, When
-                var script = new Script(new string[] { }, null, new ScriptAlias[] { }, new string[] { }, new string[] { });
+                var script = new Script(new string[] { }, null, new ScriptAlias[] { }, new string[] { }, new string[] { }, new string[] { });
 
                 // Then
                 Assert.Equal(0, script.Lines.Count);
@@ -35,7 +35,7 @@ namespace Cake.Core.Tests.Unit.Scripting
             public void Should_Not_Throw_If_Aliases_Are_Null()
             {
                 // Given, When
-                var script = new Script(new string[] { }, new string[] { }, null, new string[] { }, new string[] { });
+                var script = new Script(new string[] { }, new string[] { }, null, new string[] { }, new string[] { }, new string[] { });
 
                 // Then
                 Assert.Equal(0, script.Aliases.Count);
@@ -45,7 +45,7 @@ namespace Cake.Core.Tests.Unit.Scripting
             public void Should_Not_Throw_If_Using_Alias_Directives_Are_Null()
             {
                 // Given, When
-                var script = new Script(new string[] { }, new string[] { }, new ScriptAlias[] { }, null, new string[] { });
+                var script = new Script(new string[] { }, new string[] { }, new ScriptAlias[] { }, null, new string[] { }, new string[] { });
 
                 // Then
                 Assert.Equal(0, script.UsingAliasDirectives.Count);
@@ -55,10 +55,20 @@ namespace Cake.Core.Tests.Unit.Scripting
             public void Should_Not_Throw_If_Using_Static_Directives_Are_Null()
             {
                 // Given, When
-                var script = new Script(new string[] { }, new string[] { }, new ScriptAlias[] { }, new string[] { }, null);
+                var script = new Script(new string[] { }, new string[] { }, new ScriptAlias[] { }, new string[] { }, null, new string[] { });
 
                 // Then
                 Assert.Equal(0, script.UsingStaticDirectives.Count);
+            }
+
+            [Fact]
+            public void Should_Not_Throw_If_Defines_Are_Null()
+            {
+                // Given, When
+                var script = new Script(new string[] { }, new string[] { }, new ScriptAlias[] { }, new string[] { }, new string[] { }, null);
+
+                // Then
+                Assert.Equal(0, script.Defines.Count);
             }
         }
     }

--- a/src/Cake.Core/Scripting/Analysis/IScriptInformation.cs
+++ b/src/Cake.Core/Scripting/Analysis/IScriptInformation.cs
@@ -50,6 +50,12 @@ namespace Cake.Core.Scripting.Analysis
         IList<string> UsingStaticDirectives { get; }
 
         /// <summary>
+        /// Gets the defines directives defined by the scripts.
+        /// </summary>
+        /// <value>The defines.</value>
+        IList<string> Defines { get; }
+
+        /// <summary>
         /// Gets the tools.
         /// </summary>
         /// <value>The tools.</value>

--- a/src/Cake.Core/Scripting/Analysis/ScriptAnalyzer.cs
+++ b/src/Cake.Core/Scripting/Analysis/ScriptAnalyzer.cs
@@ -63,7 +63,8 @@ namespace Cake.Core.Scripting.Analysis
                 new AddInDirectiveProcessor(),
                 new ToolDirectiveProcessor(),
                 new ShebangProcessor(),
-                new BreakDirectiveProcessor()
+                new BreakDirectiveProcessor(),
+                new DefineDirectiveProcessor()
             };
         }
 

--- a/src/Cake.Core/Scripting/Analysis/ScriptAnalyzerResult.cs
+++ b/src/Cake.Core/Scripting/Analysis/ScriptAnalyzerResult.cs
@@ -17,6 +17,7 @@ namespace Cake.Core.Scripting.Analysis
         private readonly HashSet<string> _namespaces;
         private readonly HashSet<string> _usingAliases;
         private readonly HashSet<string> _usingStaticDirectives;
+        private readonly HashSet<string> _defines;
 
         /// <summary>
         /// Gets the analyzed script.
@@ -54,6 +55,12 @@ namespace Cake.Core.Scripting.Analysis
         public ISet<string> UsingStaticDirectives => _usingStaticDirectives;
 
         /// <summary>
+        /// Gets the defines.
+        /// </summary>
+        /// <value>The defines.</value>
+        public ISet<string> Defines => _defines;
+
+        /// <summary>
         /// Gets the addins.
         /// </summary>
         /// <value>The addins.</value>
@@ -89,6 +96,7 @@ namespace Cake.Core.Scripting.Analysis
             _namespaces = new HashSet<string>(Collect(script, i => i.Namespaces));
             _usingAliases = new HashSet<string>(Collect(script, i => i.UsingAliases));
             _usingStaticDirectives = new HashSet<string>(Collect(script, i => i.UsingStaticDirectives));
+            _defines = new HashSet<string>(Collect(script, i => i.Defines));
             Tools = new HashSet<PackageReference>(Collect(script, i => i.Tools));
             Addins = new HashSet<PackageReference>(Collect(script, i => i.Addins));
             Errors = errors ?? new List<ScriptAnalyzerError>(0);

--- a/src/Cake.Core/Scripting/Analysis/ScriptInformation.cs
+++ b/src/Cake.Core/Scripting/Analysis/ScriptInformation.cs
@@ -15,6 +15,7 @@ namespace Cake.Core.Scripting.Analysis
         private readonly List<string> _usingAliases;
         private readonly List<string> _namespaces;
         private readonly List<string> _usingStaticDirectives;
+        private readonly List<string> _defines;
         private readonly List<PackageReference> _addins;
         private readonly List<PackageReference> _tools;
         private readonly List<string> _lines;
@@ -31,6 +32,8 @@ namespace Cake.Core.Scripting.Analysis
 
         public IList<string> UsingStaticDirectives => _usingStaticDirectives;
 
+        public IList<string> Defines => _defines;
+
         public IList<PackageReference> Addins => _addins;
 
         public IList<PackageReference> Tools => _tools;
@@ -45,6 +48,7 @@ namespace Cake.Core.Scripting.Analysis
             _namespaces = new List<string>();
             _usingAliases = new List<string>();
             _usingStaticDirectives = new List<string>();
+            _defines = new List<string>();
             _addins = new List<PackageReference>();
             _tools = new List<PackageReference>();
             _lines = new List<string>();

--- a/src/Cake.Core/Scripting/Processors/DefineDirectiveProcessor.cs
+++ b/src/Cake.Core/Scripting/Processors/DefineDirectiveProcessor.cs
@@ -1,0 +1,31 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Linq;
+using Cake.Core.Scripting.Analysis;
+
+namespace Cake.Core.Scripting.Processors
+{
+    internal sealed class DefineDirectiveProcessor : LineProcessor
+    {
+        public override bool Process(IScriptAnalyzerContext context, string line, out string replacement)
+        {
+            if (context == null)
+            {
+                throw new ArgumentNullException(nameof(context));
+            }
+
+            replacement = null;
+
+            if (!line.StartsWith("#define", StringComparison.Ordinal))
+            {
+                return false;
+            }
+
+            context.Current.Defines.Add(line);
+            return true;
+        }
+    }
+}

--- a/src/Cake.Core/Scripting/Script.cs
+++ b/src/Cake.Core/Scripting/Script.cs
@@ -17,6 +17,7 @@ namespace Cake.Core.Scripting
         private readonly List<ScriptAlias> _aliases;
         private readonly List<string> _usingAliasDirectives;
         private readonly List<string> _usingStaticDirectives;
+        private readonly List<string> _defines;
 
         /// <summary>
         /// Gets the namespaces imported via the <c>using</c> statement.
@@ -51,6 +52,12 @@ namespace Cake.Core.Scripting
         public IReadOnlyList<string> UsingStaticDirectives => _usingStaticDirectives;
 
         /// <summary>
+        /// Gets the defines.
+        /// </summary>
+        /// <value>The defines.</value>
+        public IReadOnlyList<string> Defines => _defines;
+
+        /// <summary>
         /// Initializes a new instance of the <see cref="Script" /> class.
         /// </summary>
         /// <param name="namespaces">The namespaces.</param>
@@ -58,18 +65,21 @@ namespace Cake.Core.Scripting
         /// <param name="aliases">The script aliases.</param>
         /// <param name="usingAliasDirectives">The using alias directives.</param>
         /// <param name="usingStaticDirectives">The using static directives.</param>
+        /// <param name="defines">The defines.</param>
         public Script(
             IEnumerable<string> namespaces,
             IEnumerable<string> lines,
             IEnumerable<ScriptAlias> aliases,
             IEnumerable<string> usingAliasDirectives,
-            IEnumerable<string> usingStaticDirectives)
+            IEnumerable<string> usingStaticDirectives,
+            IEnumerable<string> defines)
         {
             _namespaces = new List<string>(namespaces ?? Enumerable.Empty<string>());
             _lines = new List<string>(lines ?? Enumerable.Empty<string>());
             _aliases = new List<ScriptAlias>(aliases ?? Enumerable.Empty<ScriptAlias>());
             _usingAliasDirectives = new List<string>(usingAliasDirectives ?? Enumerable.Empty<string>());
             _usingStaticDirectives = new List<string>(usingStaticDirectives ?? Enumerable.Empty<string>());
+            _defines = new List<string>(defines ?? Enumerable.Empty<string>());
         }
     }
 }

--- a/src/Cake.Core/Scripting/ScriptRunner.cs
+++ b/src/Cake.Core/Scripting/ScriptRunner.cs
@@ -207,7 +207,7 @@ namespace Cake.Core.Scripting
             }
 
             // Execute the script.
-            var script = new Script(result.Namespaces, result.Lines, aliases, result.UsingAliases, result.UsingStaticDirectives);
+            var script = new Script(result.Namespaces, result.Lines, aliases, result.UsingAliases, result.UsingStaticDirectives, result.Defines);
             session.Execute(script);
         }
 

--- a/src/Cake/Scripting/Roslyn/RoslynCodeGenerator.cs
+++ b/src/Cake/Scripting/Roslyn/RoslynCodeGenerator.cs
@@ -12,11 +12,12 @@ namespace Cake.Scripting.Roslyn
     {
         public string Generate(Script script)
         {
+            var defines = string.Join("\r\n", script.Defines);
             var usingStaticDirectives = string.Join("\r\n", script.UsingStaticDirectives);
             var usingDirectives = string.Join("\r\n", script.UsingAliasDirectives);
             var aliases = GetAliasCode(script);
             var code = string.Join("\r\n", script.Lines);
-            return string.Join("\r\n", usingStaticDirectives, usingDirectives, aliases, code);
+            return string.Join("\r\n", defines, usingStaticDirectives, usingDirectives, aliases, code);
         }
 
         private static string GetAliasCode(Script context)

--- a/tests/integration/Cake.Core/Scripting/DefineDirective.cake
+++ b/tests/integration/Cake.Core/Scripting/DefineDirective.cake
@@ -1,0 +1,33 @@
+#define FOO
+#load "./../../utilities/xunit.cake"
+
+
+Task("Cake.Core.Scripting.DefineDirective.Defined")
+    .Does(() =>
+{
+    bool foo;
+#if (FOO)
+    foo = true;
+#else
+    foo = false;
+#endif
+    Assert.True(foo);
+});
+
+Task("Cake.Core.Scripting.DefineDirective.NotDefined")
+    .Does(() =>
+{
+    bool bar;
+#if (BAR)
+    bar = true;
+#else
+    bar = false;
+#endif
+    Assert.False(bar);
+});
+
+//////////////////////////////////////////////////////////////////////////////
+
+Task("Cake.Core.Scripting.DefineDirective")
+    .IsDependentOn("Cake.Core.Scripting.DefineDirective.Defined")
+    .IsDependentOn("Cake.Core.Scripting.DefineDirective.NotDefined");

--- a/tests/integration/build.cake
+++ b/tests/integration/build.cake
@@ -20,6 +20,7 @@
 #load "./Cake.Common/Tools/Cake/CakeAliases.cake"
 #load "./Cake.Common/Tools/DotNetCore/DotNetCoreAliases.cake"
 #load "./Cake.Common/Tools/NuGet/NuGetAliases.cake"
+#load "./Cake.Core/Scripting/DefineDirective.cake"
 #load "./Cake.Core/Scripting/LoadDirective.cake"
 #load "./Cake.Core/Scripting/UsingDirective.cake"
 #load "./Cake.Core/Tooling/ToolLocator.cake"
@@ -45,6 +46,7 @@ Setup(ctx =>
 //////////////////////////////////////////////////
 
 Task("Cake.Core")
+    .IsDependentOn("Cake.Core.Scripting.DefineDirective")
     .IsDependentOn("Cake.Core.Scripting.LoadDirective")
     .IsDependentOn("Cake.Core.Scripting.UsingDirective")
     .IsDependentOn("Cake.Core.Tooling.ToolLocator")

